File: kotlin/bc-envelope/src/test/kotlin/com/blockchaincommons/bcenvelope/CheckEncoding.kt
package com.blockchaincommons.bcenvelope

import com.blockchaincommons.bccomponents.DigestProvider

/**
 * Extension for round-trip CBOR encoding verification in tests.
 */
fun Envelope.checkEncoding(): Envelope {
    val cbor = this.taggedCbor()
    val restored = try {
        Envelope.fromTaggedCbor(cbor)
    } catch (e: Exception) {
        println("=== EXPECTED")
        println(this.format())
        println("=== GOT")
        println(cbor.diagnostic())
        println("===")
        throw EnvelopeException.InvalidFormat()
    }
    if (this.digest() != restored.digest()) {
        println("=== EXPECTED")
        println(this.format())
        println("=== GOT")
        println(restored.format())
        println("===")
        throw EnvelopeException.General("Digest mismatch")
    }
    return this
}

File: python/bc-envelope/tests/common/check_encoding.py
"""CBOR round-trip verification utility for Envelope tests."""


def check_encoding(envelope):
    """Verify CBOR encoding round-trips correctly.

    Encodes the envelope to tagged CBOR, decodes it back, and asserts that
    the digest is unchanged.  Returns the original envelope on success.
    """
    from bc_envelope import Envelope

    cbor = envelope.tagged_cbor()
    restored = Envelope.from_tagged_cbor(cbor)
    assert envelope.digest() == restored.digest(), (
        f"Digest mismatch after round-trip:\n"
        f"  original: {envelope.format()}\n"
        f"  decoded:  {restored.format()}"
    )
    return envelope
